### DISCUSSION TITLE
Recommendations on how to build modules for existing native libraries

### DISCUSSION THREAD

**Discussion by bralax** [reactwg/react-native-new-architecture#170] [Q&A]
I am currently working on a project where I want to take an existing library (in my case [wgpu-native](https://github.com/gfx-rs/wgpu-native) a c binding for webgpu) and make it available to my react-native application. Looking through the docs in this wiki, the focus definitely seems to be on kind of the reverse situation of what I'm dealing with where flow/typescript is the "source of truth" for what you are trying to build and then the C++/Objective-C/Java code is just for implementation rather than having existing C++/Native code and wanting to do the reverse. From my reading so far, I have a couple of question, requests for thoughts: 
1. When working with an existing native library what is recommendation for thinking about how to generate the typescript/flow interface?
2. How can you minimize "glue code" needed to integrate typescript->jsi->bindings->library?
3. Most large packages such as the one I am looking at, have their own build systems. What's the working groups thoughts on how to include and compile these dependencies and ensure this native code Is properly compiled for the specific platform?
4. Looking at how JSII handles type conversion, there seems to be a lot of data copying (where you have to write manual converters to take a C/C++/Other native language type and convert it field by field to a javascript compatible type). This is definitely faster than the old string-based bridge but still introduces overhead. Especially when working with native resources, is there any guidance or mechanism to work with pointers? If for example I wanted to work with a file descriptor, how would I describe that type and pass the file descriptor pointer back and forth between js/C++? Is it allowed in JSII to pass pointers back and forth and if so how?

**Comment by cortinico** [comment]
> When working with an existing native library what is recommendation for thinking about how to generate the typescript/flow interface?

The recommended approach would be to create a JS/TS interface that models the same API that you get exposed from Native. That's particularly important if that same SDK needs to work between Android/iOS and other platforms.

> How can you minimize "glue code" needed to integrate typescript->jsi->bindings->library?

The reccomendation here is to use `@react-native/codegen` which takes care of generating most of the boilerplate for you.

> Most large packages such as the one I am looking at, have their own build systems. What's the working groups thoughts on how to include and compile these dependencies and ensure this native code Is properly compiled for the specific platform?

This could get really complicated. Ideally they'll build with Gradle/CMake/CocoaPods which are industry standards. If they do not, you should consider creating the build logic for Gradle/CMake/CocoaPods for the SDK, as it would get really hard to integrate with other RN projects otherwise.
